src/Microsoft.ComponentDetection.Detectors/rust/RustCrateDetector.cs

```diff
@@ -34,61 +36,207 @@ protected override Task OnFileFound(ProcessRequest processRequest, IDictionary<s
-                if (!cargoTomlComponentStream.Any() || cargoTomlComponentStream.Count() > 1)
+                if (childPackage == null)
                 {
-                    Logger.LogWarning($"We are expecting exactly 1 accompanying Cargo.toml file next to the cargo.lock file found at {cargoLockFile.Location}");
-                    return Task.CompletedTask;
+                    throw new FormatException($"Could not find matching package for dependency string '{dependency}'");
                 }
 
-                // If there is a mismatch between the number of expected and found workspaces, exit
-                if (expectedWorkspaceTomlCount > numWorkspaceComponentStreams)
+                if (IsLocalPackage(childPackage))
                 {
-                    Logger.LogWarning($"We are expecting at least {expectedWorkspaceTomlCount} accompanying Cargo.toml file(s) from workspaces outside of the root directory {lockFileInfo.DirectoryName}, but found {numWorkspaceComponentStreams}");
-                    return Task.CompletedTask;
+                    if (!IsLocalPackage(parentPackage))
+                    {
+                        throw new FormatException($"In package with source '{parentComponent.Id}' found non-source dependency string: '{dependency}'");
+                    }
+
+                    // This is a dependency between packages without source
+                    return;
                 }
 
-                var cargoPackages = cargoLock.package.ToHashSet();
-                RustCrateUtilities.BuildGraph(cargoPackages, cargoDependencyData.NonDevDependencies, cargoDependencyData.DevDependencies, singleFileComponentRecorder);
+                var detectedComponent = new DetectedComponent(childComponent);
+                seenAsDependency.Add(childPackage);
+
+                if (IsLocalPackage(parentPackage))
+                {
+                    // We are adding a root edge (from a local package)
+                    singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: true);
+                }
+                else
+                {
+                    // we are adding an edge within the graph
+                    singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: false, parentComponentId: parentComponent.Id);
+                }
             }
             catch (Exception e)
             {
-                // If something went wrong, just ignore the file
+                using var record = new RustCrateDetectorTelemetryRecord();
+
+                record.PackageInfo = $"{parentPackage.name}, {parentPackage.version}, {parentPackage.source}";
+                record.Dependencies = dependency;
+
                 Logger.LogFailedReadingFile(cargoLockFile.Location, e);
             }
+        }
 
-            return Task.CompletedTask;
+        private static readonly Regex DependencyFormatRegex = new Regex(
```

Comment:
Should this regex include start/end markers `^$` to prevent partially matching a string?

Reply:
Added